packages/daemon/daemon-docs.md

```diff
@@ -0,0 +1,112 @@
+- The powers accessible within this new context are:
+
+  - `crypto`
+  - `net`
+  - `fs`
+  - `path`
+  - `popen`
+  - `url`
+
+- takes one argument, `locator`. If nothing is passed in, this defaults to [`defaultLocator`](#defaultlocator).
+
+### `terminate`
+
+- creates a new endo client using [makeEndoClient](#makeendoclient) for the sole purpose of gaining access to the `terminate` method.
+- Uses `terminate` in order to sever any existing connections.
+- takes one argument, `locator`. If nothing is passed in, this defaults to [`defaultLocator`](#defaultlocator).
+
+```js
+export const terminate = async (locator = defaultLocator) => {
+  const { resolve: cancel, promise: cancelled } = makePromiseKit();
+  const { getBootstrap, closed } = await makeEndoClient(
+    'harbinger',
+    locator.sockPath,
+    cancelled,
+  );
+  const bootstrap = getBootstrap();
+  await E(bootstrap)
+    .terminate()
+    .catch(() => {});
+  // @ts-expect-error zero-argument promise resolve
+  cancel();
+  await closed.catch(() => {});
+};
+```
+
+### `clean`
+
+- recursively and forcibly removes any directories that have been created throughout the lifecycle of the previous Endo daemon.
+- takes one argument, `locator`. If nothing is passed in, this defaults to [`defaultLocator`](#defaultlocator).
+
+### `stop`
+
+- terminates existing Endo daemon processess before calling [`clean`](#clean).
+- takes one argument, `locator`. If nothing is passed in, this defaults to [`defaultLocator`](#defaultlocator).
+
+### `makeEndoClient`
+
+- Used to spin up a new application within the context of an Endo daemon.
+- The processess created with makeEndoClient come pre-assembled with the ability to communicate over CapTP using Node.js-based messages.
+- Utilizes the global `net` package to create a socket connection for an Endo process. (`net.connect(sockPath)`)
+
+### defaultLocator
+
+- default `locator` object which an Endo process will be initialized with.
+- See [@endo/daemon](https://github.com/endojs/endo/blob/master/packages/daemon/index.js#L26)
+
+```js
+const defaultLocator = {
+  statePath: whereEndoState(process.platform, process.env, info),
+  ephemeralStatePath: whereEndoEphemeralState(
+    process.platform,
+    process.env,
+    info,
+  ),
+  sockPath: whereEndoSock(process.platform, process.env, info),
+  cachePath: whereEndoCache(process.platform, process.env, info),
+};
+```
+
+# Questions
+
+- the `main` function seems is where “all of it” starts.
+- where does `main0` come from?
+- Regarding the locator...
+  - what is the difference between:
+    - statePath
+    - ephemeralStatPath
+    - sockPath
+    - cachePath
+  - after coming across the term "noncelocator" a number of times while reading erights.org, I'm wondering what is the locator within the context of SwingSet?
+- What is the significance of `child.on` within the endo daemon's `start` function? What exactly occurs when a “message” event occurs?
```

Comment:
This is one of the wonders of daemonization. Daemonization is the process of spawning a child process, removing it from the parent process’s session, and handing it over to process 0 (init) to manage until it dies. On UNIX, by default, a child process remains in the parent process’s session, sees all of the same signals, and exits when the parent exits. Daemons are longer-lived.

When a client like the CLI attempts to open a connection to the Endo daemon, there’s a possibility that the daemon is not running. So, the client will spawn and daemonize the daemon. But, owing to the vagaries of timing, the daemon will not be ready to receive connections from the client until it has run a while and started listening.

This is a problem for the client which wants to connect to the daemon the moment it becomes available. We’re using the Node.js IPC port to send a message from the child process (daemon) to the parent process (cli) that it can connect.

The alternative is a retry loop, which will always be slower than this approach.